sailfishusermanager.cpp

```diff
@@ -297,13 +306,42 @@ bool SailfishUserManager::removeHome(uint uid)
     return removeDir(home);
 }
 
+/*
+ * Gets caller uid and checks it has proper rights.
+ * Returns with the uid if ok, otherwise SAILFISH_UDEFINED_UID.
+ */
+uid_t SailfishUserManager::checkCallerUid()
+{
+    // Get the PID of the calling process
+    pid_t pid = connection().interface()->servicePid(message().service());
+
+    // The /proc/<pid> directory is owned by EUID:EGID of the process
+    QFileInfo info(QString("/proc/%1").arg(pid));
+    uid_t uid = info.ownerId();
+
+    if (uid == 0) {
+        // Root is always allowed to make changes
+        return uid;
+    }
+
+    if (info.group() != QStringLiteral("privileged")) {
+        // Non-privileged applications are not allowed
+        auto message = QStringLiteral("PID %1 is not in privileged group").arg(pid);
+        qCWarning(lcSUM) << "Access denied:" << message;
+        sendErrorReply(QDBusError::AccessDenied, message);
+        return SAILFISH_UNDEFINED_UID;
+    }
+
+    return uid;
+}
+
 /* Check that calling D-Bus client is allowed to make the operation
  *
  * uid_to_modify is uid of the user that is going to be changed or removed.
  * Special value SAILFISH_UNDEFINED_UID can be used to denote non-existing user
  * that does not match to calling process' user but is in the valid range.
  */
-bool SailfishUserManager::hasAccessRights(uint uid_to_modify) {
+bool SailfishUserManager::checkAccessRights(uint uid_to_modify) {
     if (!calledFromDBus()) {
```

Comment:
If you go with this approach that there are two checking methods, then I suggest that you move this to the checkCallerUid method. The point of this check is to make sure that the QDBusContext methods are available.

The alternative could be to modify checkAccessRights so that it can work with switchUser too. Either redefine the use of SAILFISH_UNDEFINED_UID or add another value, I think. (Policy blocks addUser from other users anyway.) But either approach works.

Reply:
Moved